src/scan/heap_reader_worker.cpp

```diff
@@ -0,0 +1,143 @@
+#include "tcop/tcopprot.h"
+}
+
+#include "pgduckdb/scan/heap_reader_worker.hpp"
+
+extern "C" {
+
+void
+PGDuckDBHeapReaderWorker(Datum arg) {
+	pqsignal(SIGTERM, die);
+	dsm_segment *thread_worker_segment = nullptr, *worker_shared_state_segment = nullptr;
+	BufferAccessStrategy strategy = GetAccessStrategy(BAS_BULKREAD);
+
+	BackgroundWorkerUnblockSignals();
+
+	pgduckdb::PostgresScanThreadWorker *thread_worker_shared_state = NULL;
+	pgduckdb::PostgresScanWorkerSharedState *worker_shared_state = NULL;
+
+	thread_worker_segment = dsm_attach(arg);
+	thread_worker_shared_state = (pgduckdb::PostgresScanThreadWorker *)dsm_segment_address(thread_worker_segment);
+
+	worker_shared_state_segment = dsm_attach(thread_worker_shared_state->workers_global_shared_state);
+	worker_shared_state = (pgduckdb::PostgresScanWorkerSharedState *)dsm_segment_address(worker_shared_state_segment);
+
+	BackgroundWorkerInitializeConnectionByOid(worker_shared_state->database, InvalidOid, 0);
+	StartTransactionCommand();
+
+	Relation relation = RelationIdGetRelation(worker_shared_state->relid);
+	bool thread_running = true;
+
+	while (thread_running) {
+
+		SpinLockAcquire(&worker_shared_state->worker_lock);
+
+		/* Nothing to scan next so finish this worker */
+		if (worker_shared_state->next_scan_start_block == worker_shared_state->n_blocks) {
+			SpinLockRelease(&worker_shared_state->worker_lock);
+			break;
+		}
+
+		uint32 worker_start_scan_block = worker_shared_state->next_scan_start_block;
+		uint32 worker_end_scan_block = worker_start_scan_block + pgduckdb::WORKER_SCAN_BLOCK;
+
+		if (worker_end_scan_block > worker_shared_state->n_blocks)
+			worker_end_scan_block = worker_shared_state->n_blocks;
+
+		worker_shared_state->next_scan_start_block = worker_end_scan_block;
+
+		SpinLockRelease(&worker_shared_state->worker_lock);
+
+		/* Let's fetch */
+		for (int i = worker_start_scan_block; i < worker_end_scan_block; i++) {
+
+			if (!thread_running) {
+				break;
+			}
+
+			Buffer buffer = ReadBufferExtended(relation, MAIN_FORKNUM, i, RBM_NORMAL, strategy);
+			LockBuffer(buffer, BUFFER_LOCK_SHARE);
+
+			/* is previous buffer done */
+			while (!pg_atomic_unlocked_test_flag(&thread_worker_shared_state->buffer_ready)) {
+				if (pg_atomic_unlocked_test_flag(&thread_worker_shared_state->thread_running)) {
+					thread_running = false;
+					break;
+				}
+			}
+
+			/* Thread exited so we need do cleanup current buffer */
+			if (!thread_running) {
+				UnlockReleaseBuffer(buffer);
+			}
+
+			Buffer previous_buffer = thread_worker_shared_state->buffer;
+
+			thread_worker_shared_state->buffer = buffer;
+			pg_atomic_test_set_flag(&thread_worker_shared_state->buffer_ready);
+
+			if (BufferIsValid(previous_buffer)) {
+				UnlockReleaseBuffer(previous_buffer);
+			}
+		}
+	}
+
+	if (thread_running) {
+		/* We are out of blocks fo reading so wait for last buffer to be done */
+		while (!pg_atomic_unlocked_test_flag(&thread_worker_shared_state->buffer_ready)) {
```

Comment:
I guess same apply here - and we could factorize the two loops in a "wait" function?